File: src/main/java/com/gitlab/lae/intellij/actions/tree/Popup.java
package com.gitlab.lae.intellij.actions.tree;

import com.gitlab.lae.intellij.actions.tree.ui.ActionList;
import com.gitlab.lae.intellij.actions.tree.ui.ActionPopupEventDispatcher;
import com.gitlab.lae.intellij.actions.tree.ui.ActionPresentation;
import com.intellij.ide.DataManager;
import com.intellij.ide.IdePopupManager;
import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.ui.popup.JBPopup;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.util.Consumer;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.util.List;

import static com.gitlab.lae.intellij.actions.tree.util.JBPopups.setBestLocation;
import static com.intellij.openapi.actionSystem.CommonDataKeys.EDITOR;
import static com.intellij.openapi.actionSystem.IdeActions.ACTION_EDITOR_ESCAPE;
import static com.intellij.openapi.actionSystem.PlatformDataKeys.CONTEXT_COMPONENT;
import static com.intellij.openapi.actionSystem.ex.ActionUtil.lastUpdateAndCheckDumb;
import static com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAware;
import static com.intellij.openapi.ui.popup.JBPopupFactory.ActionSelectionAid.NUMBERING;
import static com.intellij.openapi.util.AsyncResult.done;
import static java.util.Objects.requireNonNull;
import static java.util.stream.Collectors.toList;

final class Popup {

    private final ActionManager actionManager;
    private final Component sourceComponent;
    private final Editor sourceEditor;
    private final IdePopupManager popupManager;
    private final JBPopupFactory popupFactory;
    private final DataManager dataManager;
    private final ActionList list;
    private final JBPopup popup;

    Popup(
            ActionNode action,
            AnActionEvent e,
            IdePopupManager popupManager,
            JBPopupFactory popupFactory,
            DataManager dataManager
    ) {
        this.popupManager = requireNonNull(popupManager);
        this.popupFactory = requireNonNull(popupFactory);
        this.dataManager = requireNonNull(dataManager);
        this.actionManager = e.getActionManager();
        this.sourceComponent = e.getData(CONTEXT_COMPONENT);
        this.sourceEditor = e.getData(EDITOR);

        List<ActionPresentation> items = action.items().stream()
                .map(it -> createPresentation(it, e))
                .collect(toList());

        list = new ActionList(items);

        // Register our action first before IntelliJ registers the default
        // actions (e.g. com.intellij.ui.ScrollingUtil) so that in case of
        // conflict our action will be executed
        items.forEach(item -> item.registerShortcuts(
                list,
                this::onActionChosen
        ));

        popup = createPopup();
        popup.addListener(new ActionPopupEventDispatcher(
                popup,
                list,
                popupManager
        ));

        registerIdeAction(ACTION_EDITOR_ESCAPE, popup::cancel);
    }

    void show(DataContext dataContext) {
        popup.showInBestPositionFor(dataContext);
    }

    private ActionPresentation createPresentation(
            ActionNode action,
            AnActionEvent event
    ) {
        return action.createPresentation(
                event,
                popupManager,
                popupFactory,
                dataManager
        );
    }

    private JBPopup createPopup() {
        return popupFactory
                .createListPopupBuilder(list)
                .setModalContext(true)
                .setItemChoosenCallback(() -> {
                    ActionPresentation item = list.getSelectedValue();
                    if (item != null) {
                        onActionChosen(item, 0);
                    }
                })
                .createPopup();
    }

    private void registerIdeAction(
            String actionId,
            Runnable runnable
    ) {
        AnAction action = actionManager.getAction(actionId);
        if (action == null) {
            return;
        }
        ShortcutSet shortcutSet = action.getShortcutSet();
        if (shortcutSet.getShortcuts().length == 0) {
            return;
        }
        new AnAction() {
            @Override
            public void actionPerformed(AnActionEvent e) {
                runnable.run();
            }
        }.registerCustomShortcutSet(shortcutSet, list);
    }

    private void onActionChosen(ActionPresentation item, ActionEvent e) {
        onActionChosen(item, e.getModifiers());
    }

    private void onActionChosen(ActionPresentation item, int modifiers) {
        if (!item.presentation().isEnabled()) {
            return;
        }

        list.setSelectedValue(item, false);

        Runnable invocation = () -> performAction(item.action(), modifiers);

        if (item.sticky()) {
            invocation.run();
            updatePopupLocation();
        } else {
            popup.setFinalRunnable(invocation);
            popup.closeOk(null);
        }
    }

    private void updatePopupLocation() {
        if (sourceEditor != null) {
            setBestLocation(popup, popupFactory, sourceEditor);
        } else if (sourceComponent instanceof JComponent) {
            setBestLocation(
                    popup,
                    popupFactory,
                    (JComponent) sourceComponent
            );
        }
    }

    /* Use ACTION_SEARCH as the action place seems to work the best.
     *
     * 'Run | Stop' menu action works correctly this way by
     * showing a list of processes to stop
     *
     * 'Exit' actions works (doesn't work if place is MAIN_MENU)
     */
    private static final String ACTION_PLACE =
            ActionPlaces.ACTION_SEARCH;

    private void performAction(AnAction action, int modifiers) {
        (sourceComponent == null
                ? dataManager.getDataContextFromFocus()
                : done(dataManager.getDataContext(sourceComponent))
        ).doWhenDone((Consumer<DataContext>) dataContext -> performAction(
                action,
                modifiers,
                dataContext
        ));
    }

    private void performAction(
            AnAction action,
            int modifiers,
            DataContext dataContext
    ) {
        AnActionEvent event = new AnActionEvent(
                null,
                dataContext,
                ACTION_PLACE,
                action.getTemplatePresentation().clone(),
                actionManager,
                modifiers
        );
        event.setInjectedContext(action.isInInjectedContext());

        if (showPopupIfGroup(action, event)) {
            return;
        }
        if (lastUpdateAndCheckDumb(action, event, false)) {
            performActionDumbAware(action, event);
        }
    }

    private boolean showPopupIfGroup(AnAction action, AnActionEvent e) {
        if (!(action instanceof ActionGroup)) {
            return false;
        }
        ActionGroup group = (ActionGroup) action;
        if (group.canBePerformed(e.getDataContext())) {
            return false;
        }
        popupFactory.createActionGroupPopup(
                e.getPresentation().getText(),
                group,
                e.getDataContext(),
                NUMBERING,
                true
        ).showInBestPositionFor(e.getDataContext());
        return true;
    }
}

File: src/main/kotlin/com/gitlab/lae/intellij/actions/tree/Popup.kt
package com.gitlab.lae.intellij.actions.tree

import com.gitlab.lae.intellij.actions.tree.ui.ActionList
import com.gitlab.lae.intellij.actions.tree.ui.ActionPopupEventDispatcher
import com.gitlab.lae.intellij.actions.tree.ui.ActionPresentation
import com.gitlab.lae.intellij.actions.tree.util.setBestLocation
import com.intellij.ide.DataManager
import com.intellij.ide.IdeEventQueue
import com.intellij.openapi.actionSystem.ActionGroup
import com.intellij.openapi.actionSystem.ActionManager
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys.EDITOR
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.actionSystem.IdeActions.ACTION_EDITOR_ESCAPE
import com.intellij.openapi.actionSystem.PlatformDataKeys.CONTEXT_COMPONENT
import com.intellij.openapi.actionSystem.ex.ActionUtil
import com.intellij.openapi.application.ApplicationManager.getApplication
import com.intellij.openapi.ui.popup.JBPopup
import com.intellij.openapi.ui.popup.JBPopupFactory
import com.intellij.openapi.ui.popup.JBPopupFactory.ActionSelectionAid
import com.intellij.openapi.ui.popup.PopupChooserBuilder
import com.intellij.openapi.wm.IdeFocusManager
import java.awt.event.ActionEvent
import javax.swing.JComponent
import javax.swing.KeyStroke
import org.jetbrains.concurrency.resolvedPromise

internal class Popup(action: ActionNode, e: AnActionEvent) {

  private val actionManager = e.actionManager
  private val sourceComponent = e.getData(CONTEXT_COMPONENT)
  private val sourceEditor = e.getData(EDITOR)
  private val list: ActionList
  private val popup: JBPopup

  init {
    val items =
        action.prepare(e.dataContext).map { (keys, item) ->
          createPresentation(item, e.actionManager, e.dataContext, keys)
        }
    list = ActionList(items)

    // Register our action first before IntelliJ registers the default
    // actions (e.g. com.intellij.ui.ScrollingUtil) so that in case of
    // conflict our action will be executed
    items.forEach { it.registerShortcuts(list) { item, e -> this.onActionChosen(item, e) } }
    popup = createPopup()
    popup.addListener(
        ActionPopupEventDispatcher(
            popup,
            list,
            IdeEventQueue.getInstance().popupManager,
        ),
    )
    registerIdeAction(ACTION_EDITOR_ESCAPE, popup::cancel)
  }

  fun show(dataContext: DataContext) {
    popup.showInBestPositionFor(dataContext)
  }

  private fun createPresentation(
      action: ActionNode,
      actionManager: ActionManager,
      dataContext: DataContext,
      keysOverride: List<KeyStroke>,
  ): ActionPresentation =
      action.createPresentation(
          actionManager,
          dataContext,
          keysOverride,
      )

  private fun createPopup(): JBPopup =
      PopupChooserBuilder(list)
          .setModalContext(true)
          .setCloseOnEnter(false)
          .setItemChoosenCallback {
            val item = list.selectedValue
            if (item != null) {
              onActionChosen(item, 0)
            }
          }
          .createPopup()

  private fun registerIdeAction(actionId: String, runnable: () -> Unit) {
    val action = actionManager.getAction(actionId) ?: return
    val shortcutSet = action.shortcutSet
    if (shortcutSet.shortcuts.isEmpty()) {
      return
    }

    object : AnAction() {
          override fun actionPerformed(e: AnActionEvent) {
            runnable()
          }
        }
        .registerCustomShortcutSet(shortcutSet, list)
  }

  private fun onActionChosen(item: ActionPresentation, e: ActionEvent) {
    onActionChosen(item, e.modifiers)
  }

  private fun onActionChosen(item: ActionPresentation, modifiers: Int) {
    if (!item.presentation.isEnabled) {
      return
    }

    list.setSelectedValue(item, false)
    val invocation = { performAction(item.action, modifiers) }
    if (item.sticky) {
      invocation()
      getApplication().invokeLater {
        updatePopupLocation()
        updatePresentations()
      }
    } else {
      popup.setFinalRunnable(invocation)
      popup.closeOk(null)
    }
  }

  private fun updatePopupLocation() {
    val popupFactory = JBPopupFactory.getInstance()
    if (sourceEditor != null) {
      popup.setBestLocation(popupFactory, sourceEditor)
    } else if (sourceComponent is JComponent) {
      popup.setBestLocation(popupFactory, sourceComponent)
    }
  }

  /**
   * Updates list items to show correct information regarding current state, such as whether an
   * action should be enabled/disabled for current cursor position.
   */
  private fun updatePresentations() {
    getDataContextAsync { context ->
      generateSequence(0, Int::inc).take(list.model.size).map(list.model::getElementAt).forEach {
        it.update(actionManager, context)
      }
    }
  }

  private fun getDataContextAsync(consumer: (DataContext) -> Unit) {
    val dataManager = DataManager.getInstance()
    (if (sourceComponent == null) {
          dataManager.dataContextFromFocusAsync
        } else {
          resolvedPromise(dataManager.getDataContext(sourceComponent))
        })
        .then(consumer)
  }

  private fun performAction(action: AnAction, modifiers: Int) {
    /*
     * Wrapping with invokeLater()
     * then IdeFocusManager.doWhenFocusSettlesDown()
     * is required to get back to the pre-popup focus state
     * before executing the action, as some action won't
     * get executed correctly if focus is not restored, such
     * as 'Goto next Splitter'.
     */
    getDataContextAsync { context ->
      getApplication().invokeLater {
        IdeFocusManager.findInstance().doWhenFocusSettlesDown {
          // Another invokeLater() to put us back on the right thread
          getApplication().invokeLater { performAction(action, modifiers, context) }
        }
      }
    }
  }

  private fun performAction(
      action: AnAction,
      modifiers: Int,
      dataContext: DataContext,
  ) {
    val event =
        AnActionEvent(
            null,
            dataContext,
            ActionNode.ACTION_PLACE,
            action.templatePresentation.clone(),
            actionManager,
            modifiers,
        )

    event.setInjectedContext(action.isInInjectedContext)
    if (showPopupIfGroup(action, event)) {
      return
    }

    if (ActionUtil.lastUpdateAndCheckDumb(action, event, false)) {
      ActionUtil.performActionDumbAwareWithCallbacks(action, event)
    }
  }

  private fun showPopupIfGroup(action: AnAction, e: AnActionEvent): Boolean {
    if (action !is ActionGroup) {
      return false
    }
    if (e.presentation.isPerformGroup) {
      return false
    }

    JBPopupFactory.getInstance()
        .createActionGroupPopup(
            e.presentation.text,
            action,
            e.dataContext,
            ActionSelectionAid.NUMBERING,
            true,
        )
        .showInBestPositionFor(e.dataContext)

    return true
  }
}
